oeps/oep-0017-feature-toggles.rst

```diff
@@ -0,0 +1,809 @@
+only, think iteratively. That is, instead of having a grand toggle to unlock many
+changes at once, consider breaking up your feature into `iterative verticals`_ that
+can be released (enabled) a bit at a time. See `Release Toggles Are The Last Thing
+You Should Do`_.
+
+.. _iterative verticals: https://watirmelon.blog/2015/02/02/iterative-vs-incremental-software-development/
+.. _Release Toggles Are The Last Thing You Should Do: https://martinfowler.com/bliki/FeatureToggle.html#ReleaseTogglesAreTheLastThingYouShouldDo
+
+Use Case 2: Launch Date
+~~~~~~~~~~~~~~~~~~~~~~~
+.. _Launch Date: `Use Case 2: Launch Date`_
+
+There may be a business case, advocated by the product team, to use a toggle to
+expose a new feature on a specific grand opening date. However, for a confident
+unveiling, this use case should be used in consideration with Ops_ and/or `Beta
+Testing`_ scenarios.
+
+Ops
+~~~
+
+Dynamically controlling feature toggles, without needing to re-deploy an application,
+comes in very handy when considering the operational requirements for uptime metrics.
+
+This use case is usually driven by the engineering team.
+
+Use Case 3: Ops - Monitored Rollout
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. _Ops - Monitored Rollout: `Use Case 3: Ops - Monitored Rollout`_
+
+As teams balance the needs for rapid agile development while continuously deploying
+to a large-scale system with 99.99% uptime requirements, they need the ability to
+test new changes in production while having the ability to revert quickly. That is,
+moving rapidly and taking risks can decrease Mean Time to Failure (MTTF), which needs
+to be counterbalanced with the ability to reduce `Mean Time to Recovery (MTTR)`_.
+
+When a team is concerned about potential performance or scalability issues with an
+upcoming change, gating the change behind a toggle allows the team to:
+
+* control when the change is enabled so they can monitor it in production at their
+  own time, independent of the deployment cycle
+
+* quickly disable the change in case of unexpected issues in production
+
+* gradually rollout the change (`canary release`_), starting with a small percentage
+  of random users, detecting regressions, addressing any issues that arise, before
+  enabling for everyone.
+
+Once the team is confident about their change and the change is released to all users,
+they would safely remove the gating toggle.
+
+.. _Mean Time to Recovery (MTTR): https://www.thoughtworks.com/radar/techniques/focus-on-mean-time-to-recovery
+.. _canary release: https://martinfowler.com/bliki/CanaryRelease.html
+
+Use Case 4: Ops - Graceful Degradation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. _Ops - Graceful Degradation: `Use Case 4: Ops - Graceful Degradation`_
+
+In certain cases, the development team (in consultation with the operations team)
+may choose to extend the lifetime of an Ops toggle in the codebase even after
+releasing its gated feature. A small number of such long-lived Ops "kill switches"
```

Comment:
Do we want any best practice around naming conventions for this or other long-lived switches so they are easier to reason about?

Reply:
Agreed.  In the report section below, I proposed having developers specify "Current Use Case(s)" and "All Use Case(s)" for each of their toggles.  This should indicate whether the toggle is long-term or short-term.  I suggest keeping it decoupled from the name itself - since the use case of a toggle tends to evolve as it is used.